### ISSUE TITLE
Notebook: Mixing markdown and ggplot in for loop does not keep correct order

### ISSUE THREAD

**Issue by heseber** [rstudio/rstudio#1795]
I want to use a for loop to create a header and a graph for each element of a vector (see below). In R notebooks (*.nb.html) this works for the plain `plot` function, but not for `ggplot`. 

The attached short Rmd notebook and the HTML output show both issues - graphs created in a for loop are not shown in the right order in the nb.html output, and all headers are displayed above the first graph.

[rstudio_issue.zip](https://github.com/rstudio/rstudio/files/1513047/rstudio_issue.zip)

Please note that this bug does only occur when creating R notebook output (*.nb.html) by saving an R notebook Rmd file or using the "Preview" button, whereas the output of standard html is correct (i.e. knitr to html). Standard html output does not have any of the notebook features such as a floating table of contents or hiding/showing code chunks, therefore just switching to "knitr to html" is not an option. 

This issue has been described previously in the the commercial RStudio Support issue tracker (https://support.rstudio.com/hc/en-us/requests/23504).

**Comment by dfalty** [comment]
Verified this notebook behavior on 1.1.392 (Mac 10.12.2). There is definitely a difference between the knitting HTML output and the notebook HTML. Thanks for the report!

**Comment by Aariq** [comment]
I've found this occurs when a function inside the for loop throws a warning, but not when there are no warnings produced.  It occurs in PDF, Word, html, and Preview outputs.  Here's a reproducible example: 
[for loop problem.Rmd.zip](https://github.com/rstudio/rstudio/files/1928731/for.loop.problem.Rmd.zip)

**Comment by heseber** [comment]
Is there any news on this topic? I don't think this problem only occurs if the function inside the loop throws a warning, see my example code.

**Comment by Aariq** [comment]
You're right that these are two separate issues.  I think issue 1 has been fixed. As for for-loops that print warnings, the behavior you describe is not what happens, so there is still a bug.  If you check the reproducible example above, you should see that the order is: R console line 1, warning 1, R console line 2, plot 1, R console line 3, warning 3, plot 2, etc.  

The warnings and R output are "out of sync" with the plots by one and the second warning message never gets printed.

**Comment by heseber** [comment]
@ronblum

Thank you for the update. Right now I am still on Redhat 7 and R 3.4.3. R on our RStudio Pro server is administered by our Research IT, so I have to wait for them to upgrade to R 3.5 before I can check if this issue is indeed resolved with R 3.5. I'll send another comment once this is done.

As for the headers - right, I understand what you are saying. The issue I have is this: If I generate a larger number of plots in a for loop (say, for example, a couple of dozen plots for a couple of dozen genes), it is difficult to search for a particular plot because images are of course not searchable in the browser. That's why the idea was to add headers that would be searchable. And with "knit to html" this does work.

Do you have any idea how to address this issue?

**Comment by ronblum** [comment]
@heseber Regarding the graphs showing out of order: thanks for checking. Hopefully that resolves the problem. As for the treatment of headers when working within the notebook, previewing, etc., that's a good question. A good place to ask for advice on what you’re looking for is the RStudio Community site, especially the [R Markdown ](https://community.rstudio.com/c/R-Markdown) and [RStudio IDE ](https://community.rstudio.com/c/rstudio-ide) sections. Also, you might want to check out StackOverflow for [R-tagged questions](https://stackoverflow.com/questions/tagged/r) and suggestions.

@Aariq I see what you mean. I addition, note that when running the individual chunk, warning no. 2 is included in plot no. 1's output instead of the console output: 
<img width="614" alt="warning_2" src="https://user-images.githubusercontent.com/31009952/42514057-7740342a-8426-11e8-8c60-2113853f0c23.png">
So, there appears to be a third issue, related to the second warning (and maybe other situations) getting skipped/misbehaving. Can you please file this as a separate issue? Thanks!

**Comment by Aariq** [comment]
@heseber Have you tried playing around with the chunk option `results=markup`?

**Comment by heseber** [comment]
No, should I? 

Am 19. Juli 2018 13:19:33 MESZ schrieb Eric R Scott <notifications@github.com>:
>@heseber Have you tried playing around with the chunk option
>`results=markup`? 
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/rstudio/rstudio/issues/1795#issuecomment-406242541

-- 
Dr. Henrik Seidel, Grenobler Ring 16, 13127 Berlin
Tel. +49 30 4448650, Mobile +49 1522 8650172

**Comment by Aariq** [comment]
I haven't used it, but it sounds like it takes R output and converts it to
latex, which then gets knit. I'm not sure it would solve the out-of-order
issue, but it sounds useful for what you're trying to do—make text headers
out of R output that are then searchable.

**Comment by joelmlevin** [comment]
I'm having the same issue running 3.5.1 on Mac. None of the plots throw warnings. The order is the same in the rmd window and in the html output. Searchability is not my primary concern, I'd just like the results to appear in order for my colleagues. 

`plot_svm_cost <- function(cost_val) {
  data <- readRDS(paste0("test_data_word_", cost_val, ".rds"))
  ggplot(data, aes(mean_aggression, prediction_svm)) +
  geom_point(size=I(0.3), alpha = 0.5) +
  labs(y="SVM Prediction", x="Average Human Rating", 
       title = paste("SVM Performance:", "Cost =", cost_val), 
       subtitle = paste("Correlation =", data %$% round(cor(mean_aggression, prediction_svm), 2)))
}

costs <- c(0.25, 0.8, 1.25, 1.75, 2.5, 4, 6, 10)

for (x in costs) {
  print(plot_svm_cost(x))
}
`

**Comment by Jakob37** [comment]
Is there any news on this issue, or any workarounds? I am running into the same problems when producing a sequence of plots. As I am studying a dozen of sequential plots this becomes a quite annoying issue.

Minimal exampel. I obtain the same incorrect plot order every time I rerun the code: 9, 5, 7, 8, 1, 4, 2, 3, 6, 10.

```
for (points in seq_len(10)) {
    df <- data.frame(x=rnorm(points), y=rnorm(points))
    print(ggplot(df, aes(x, y)) + geom_point() + ggtitle(paste("Iter", points)))
}
```

**Comment by Aariq** [comment]
Well, it seems like it only happens when plots produce warnings, so any method of suppressing warnings should be a workaround.  Check out quietly() from the purrr package.  It captures the results of a function in a list. If you *want* the warning messages, you could then explicitly print the warning message followed by the plot in your for loop I think.

**Comment by Aariq** [comment]
Very odd.  I tried this and I consistently get a *different* incorrect order: 6, 7, 5, 4, 1, 3, 2, 9, 8, 10.  The other weird thing is it is not unique to a for loop.  If I use map() to create a list, the list is in the correct order, but the plots don't print in the correct order.

```{r}
test <- map(1:10,
    ~data.frame(x = rnorm(.), y = rnorm(.)) %>% 
      ggplot(aes(x, y)) + geom_point() + ggtitle(paste("Iter", .)))

test[[1]] #this is "iter 1"
test[[10]] #this is "iter 10"
test #but this prints them in the same wrong order as the for loop
```

**Comment by kevinushey** [comment]
Do you still see this issue with the [preview release](https://www.rstudio.com/products/rstudio/download/preview/)?

**Comment by Aariq** [comment]
I'm using Version 1.2.1268, so a preview release, but not the most recent.  I'll try with the most recent later.

**Comment by kevinushey** [comment]
Okay, I can reproduce there. Just to be clear, this only reproduces in R Markdown notebooks. This is an MRE:

`````
---
output: html_document
---

```{r}
for (i in 1:10)
  print(ggplot() + ggtitle(i))
```
`````

Note that the same issue does not seem to occur for base R plots, surprisingly.

**Comment by heseber** [comment]
You are right, but I think I wrote all of that in my original report. 

Am 21. Februar 2019 18:52:02 MEZ schrieb Kevin Ushey <notifications@github.com>:
>Okay, I can reproduce there. Just to be clear, this only reproduces in
>R Markdown notebooks. This is an MRE:
>
>`````
>---
>output: html_document
>---
>
>```{r}
>for (i in 1:10)
>  print(ggplot() + ggtitle(i))
>```
>`````
>
>Note that the same issue does not seem to occur for base R plots,
>surprisingly.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/rstudio/rstudio/issues/1795#issuecomment-466098399

-- 
Dr. Henrik Seidel, Grenobler Ring 16, 13127 Berlin
Tel. +49 30 4448650, Mobile +49 1522 8650172

**Comment by kevinushey** [comment]
Sorry, yes, you are correct -- I just validated what you had concluded.

**Comment by dfalty** [comment]
Verified fixed 1.2.1316-1

**Comment by vondoRishi** [comment]
I have found a workaround to this. I am using ggplot2 (3.2) + RStudio (1.2.1335) in Ubuntu 14. Although the solution is not very logical and given below


```
---
output: html_notebook
---

# Hello {.tabset}

```{r, results='asis',echo=FALSE}
library(ggplot2)
for(i in 1:5){
  cat('##',i,' \n')
  print( ggplot(iris,aes(y=Sepal.Length,x=Species))+ geom_boxplot()+ ggtitle(paste("plot",i)))
  plot.new()
  dev.off()
  cat(' \n \n')
}

```

Adding both **plot.new(),   dev.off()** solves the problem. However, it adds a small text "NA" at the _end_ of the last tab which could be ignored as of now.

**Comment by kevinushey** [comment]
The issue should no longer occur with RStudio 1.2.1335. If you're still seeing a problem, can you please share a reproducible example?

**Comment by vondoRishi** [comment]
Here is a reproducible example with **sessioninfo**
[test.nb.html.tar.gz](https://github.com/rstudio/rstudio/files/3333474/test.nb.html.tar.gz)

**Comment by kevinushey** [comment]
Thanks -- I can reproduce the issue now, and it is indeed specific to the use of `{.tabset}`s. (In fact, if you use base R plots without the plot.new() + dev.off() bit then you'll find that the base R plots end up in the wrong tabset as well)

Internal note: the issue here ultimately is that console output and plot outputs are not interleaved; all console output is collected into a single file while plot outputs are then generated 'after'.

**Comment by kevinushey** [comment]
Thinking a bit more, I'm not sure if we can realistically fix this ourselves. In theory, after plotting a ggplot2 plot, you might want to use `grid` to annotate the plot afterwards, e.g.

```
# make a ggplot2 plot
p <- ggplot(mtcars, aes(wt, mpg)) +
    geom_point()
print(p)

# draw on it afterwards
l <- linesGrob()
grid.draw(l)
```

In theory we _could_ manage the state of the graphics device after a call to `print.ggplot()` is called, but this could break some user code.

I believe this implies you do need to explicitly manage the state of the graphics device if you want to control where plots show up when interleaved with console output, as you have demonstrated in your example.

**Comment by EricGoldsmith** [comment]
[This](https://github.com/rstudio/rstudio/issues/1795#issuecomment-505779701) workaround no longer appears to work with RStudio 1.3 and ggplot 3.3.1.

**Comment by annahyan** [comment]
`result = 'asis'` fixed this for me

**Comment by AleksZakirov** [comment]
No solution to the issue?

**Comment by emugmor** [comment]
@SimonCoulombe I got the same result when I tried the `plot.new()` + `dev.off()` [workaround ](https://github.com/rstudio/rstudio/issues/1795#issuecomment-505779701)(first tab had a plot but the others were empty), but when I tried just `plot.new()` without `dev.off()` it worked great.

**Comment by avishai987** [comment]
> I have found a workaround to this. I am using ggplot2 (3.2) + RStudio (1.2.1335) in Ubuntu 14. Although the solution is not very logical and given below
> 
> ```
> ---
> output: html_notebook
> ---
> 
> # Hello {.tabset}
> 
> ```{r, results='asis',echo=FALSE}
> library(ggplot2)
> for(i in 1:5){
>   cat('##',i,' \n')
>   print( ggplot(iris,aes(y=Sepal.Length,x=Species))+ geom_boxplot()+ ggtitle(paste("plot",i)))
>   plot.new()
>   dev.off()
>   cat(' \n \n')
> }
> 
> cat('#')
> ```
> 
> Adding both **plot.new(), dev.off()** solves the problem.

This workaround still works with RStudio 2022.02.1+461for Ubuntu.
However, the issue now is that if I want to render with rmarkdown:render() the plots suddenly do not appear...
